ocaml/xapi/repository_helpers.ml

```diff
@@ -1076,3 +1076,24 @@ let prune_updateinfo_for_livepatches livepatches updateinfo =
     List.filter (fun x -> LivePatchSet.mem x livepatches) updateinfo.livepatches
   in
   {updateinfo with livepatches= lps}
+
+let do_with_device_models ~__context ?(vm_with_recommended_guidances = false)
+    ~host f =
+  (* Call f with device models of all running HVM VMs on the host *)
+  Db.Host.get_resident_VMs ~__context ~self:host
+  |> List.filter (fun self ->
+         if vm_with_recommended_guidances then
+           List.mem `restart_device_model
+             (Db.VM.get_recommended_guidances ~__context ~self)
+         else
+           true
+     )
+  |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
+  |> List.filter (fun (_, record) -> not record.API.vM_is_control_domain)
+  |> List.filter_map f
+  |> function
+  | [] ->
+      ()
+  | _ :: _ ->
+      let host' = Ref.string_of host in
+      raise Api_errors.(Server_error (cannot_restart_device_model, [host']))
```

Comment:
Only VMs have a device model - so why is the host part of the message?

Reply:
It is expected that any failure when executing "f" should be logged by "f" itself, and as the error here happens when trying to restart device model for the "host", the "cannot_restart_device_model" error of the host is logged here.